pkg/destroy/vsphere/vsphere.go

```diff
@@ -244,22 +245,59 @@ func (o *ClusterUninstaller) deleteHostZoneObjects(ctx context.Context) error {
 	return utilerrors.NewAggregate(errs)
 }
 
+func (o *ClusterUninstaller) deleteCnsVolumes(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*30)
+	defer cancel()
+	o.Logger.Debug("Delete CNS Volumes")
+
+	for _, client := range o.clients {
+		cnsVolumes, err := client.GetCnsVolumes(ctx, o.InfraID)
+		if err != nil {
+			return err
+		}
+
+		for _, cv := range cnsVolumes {
+			cnsVolumeLogger := o.Logger.WithField("CNS Volume", cv.VolumeId.Id)
+			err := client.DeleteCnsVolumes(ctx, cv)
+			if err != nil {
+				return err
+			}
+			cnsVolumeLogger.Info("Destroyed")
+		}
+	}
+
+	return nil
+}
+
 func (o *ClusterUninstaller) destroyCluster(ctx context.Context) (bool, error) {
 	stagedFuncs := [][]struct {
 		name    string
 		execute func(context.Context) error
 	}{{
-		{name: "Stop virtual machines", execute: o.stopVirtualMachines},
-	}, {
-		{name: "Virtual Machines", execute: o.deleteVirtualMachines},
-	}, {
-		{name: "Folder", execute: o.deleteFolder},
-	}, {
-		{name: "Storage Policy", execute: o.deleteStoragePolicy},
-		{name: "Tag", execute: o.deleteTag},
-		{name: "Tag Category", execute: o.deleteTagCategory},
-	}, {
-		{name: "VM Groups and VM Host Rules", execute: o.deleteHostZoneObjects},
+		{
+			name: "Stop virtual machines", execute: o.stopVirtualMachines,
+		},
+		{
+			name: "Delete Virtual Machines", execute: o.deleteVirtualMachines,
+		},
+		{
+			name: "Folder", execute: o.deleteFolder,
+		},
+		{
+			name: "Storage Policy", execute: o.deleteStoragePolicy,
+		},
+		{
+			name: "Tag", execute: o.deleteTag,
+		},
+		{
+			name: "Tag Category", execute: o.deleteTagCategory,
+		},
+		{
+			name: "VM Groups and VM Host Rules", execute: o.deleteHostZoneObjects,
+		},
+		{
+			name: "Delete CNS Volumes", execute: o.deleteCnsVolumes,
+		},
```

Comment:
Shouldn't delete CNS volumes be done *before* tags, categories or even storage policies are deleted? IIRC - if one or more volumes that refer the storage policy being deleted still exists then deletion of storage policy fails (this is one of the primary reasons we have steady accumulation of storage policies in devqe environments I think). 

But please do double check the order.

Reply:
Thanks @gnufied, let me test it